File: TracksSample/BasicSample/ViewController.m
//
//  ViewController.m
//  BasicSample
//
//  Created by Gal Orlanczyk on 15/03/2017.
//  Copyright © 2017 Kaltura. All rights reserved.
//

#import "ViewController.h"
#import "PlayKit-Swift.h"

@interface ViewController ()

@property (strong, nonatomic) id<Player> player;
@property (strong, nonatomic) NSTimer *playheadTimer;
@property (weak, nonatomic) IBOutlet PlayerView *playerContainer;
@property (weak, nonatomic) IBOutlet UISlider *playheadSlider;

@property (weak, nonatomic) NSArray *audioTracks;
@property (weak, nonatomic) NSArray *textTracks;
@property (weak, nonatomic) NSArray *selectedTracks;
    
@end

@implementation ViewController

/*********************************/
#pragma mark - LifeCycle
/*********************************/

- (void)viewDidLoad {
    [super viewDidLoad];
    self.playheadSlider.continuous = NO;
    // setup our player instance
    [self setupPlayer];
    
    // handle audio/ text tracks
    [self handleTracks];
    // get current bitrate value
    [self currentBitrateHandler];
}

/*********************************/
#pragma mark - Player Setup
/*********************************/
    
- (void)setupPlayer {
    NSURL *contentURL = [[NSURL alloc] initWithString:@"http://devimages.apple.com/iphone/samples/bipbop/bipbopall.m3u8"];
    
    // create media source and initialize a media entry with that source
    NSString *entryId = @"apple_bipbop";
    PKMediaSource* source = [[PKMediaSource alloc] init:entryId contentUrl:contentURL mimeType:nil drmData:nil mediaFormat:MediaFormatHls];
    NSArray<PKMediaSource*>* sources = [[NSArray alloc] initWithObjects:source, nil];
    // setup media entry
    PKMediaEntry *mediaEntry = [[PKMediaEntry alloc] init:entryId sources:sources duration:-1];
    
    // create media config
    MediaConfig *mediaConfig = [[MediaConfig alloc] initWithMediaEntry:mediaEntry startTime:0.0];
    
    // load the player
    NSError *error = nil;
    self.player = [PlayKitManager.sharedInstance loadPlayerWithPluginConfig:nil error:&error];
    
    if (!error) {
        [self.player prepare:mediaConfig];
        self.player.view = self.playerContainer;
        [self.playerContainer sendSubviewToBack:self.player.view];
    } else {
        // error loading the player
    }
}

/*********************************/
#pragma mark - Tracks
/*********************************/

// Handle Available Tracks and Present Them
- (void)handleTracks {
    // don't forget to use weak self to prevent retain cycles when needed
    __weak __typeof(self) weakSelf = self;
    
    // add observer to tracksAvailable event
    [self.player addObserver:self events:@[PlayerEvent.tracksAvailable] block:^(PKEvent * _Nonnull event) {
        if ([event isKindOfClass:PlayerEvent.tracksAvailable]) {
            // Connect data
            weakSelf.picker.dataSource = weakSelf;
            weakSelf.picker.delegate = weakSelf;
            
            // Extract Audio Tracks
            if (event.tracks.audioTracks) {
                weakSelf.audioTracks = event.tracks.audioTracks;
                
                // Set Defualt array for Picker View
                weakSelf.selectedTracks = weakSelf.audioTracks;
            }
            // Extract Text Tracks
            if (event.tracks.textTracks) {
                weakSelf.textTracks = event.tracks.textTracks;
            }
        }
    }];
    
    [self printCurrentTrack];
}


// Get Current Audio/ Text Track
- (void)printCurrentTrack {
    NSLog(@"currentAudioTrack:: %@", self.player.currentAudioTrack);
    NSLog(@"currentTextTrack:: %@", self.player.currentTextTrack);
}

// Get Current Bitrate
- (void)currentBitrateHandler {   
    [self.player addObserver:self events:@[PlayerEvent.playbackInfo] block:^(PKEvent * _Nonnull event) {
        if ([event isKindOfClass:PlayerEvent.playbackInfo]) {
            // Get Current Bitrate Value
            if (event.playbackInfo) {
                NSLog(@"%@", event.playbackInfo);
            }
        }
    }];
}

// Select Track
- (void)selectTrack:(Track *)track {
    [self.player selectTrackWithTrackId:track.id];
}

/*********************************/
// Tracks - Picker View
/*********************************/
- (IBAction)audioTracksTapped:(id)sender {
    self.selectedTracks = self.audioTracks;
    [self.picker reloadAllComponents];
}

- (IBAction)textTracksTapped:(id)sender {
    self.selectedTracks = self.textTracks;
    [self.picker reloadAllComponents];
}

// The number of columns of data
- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView {
    return 1;
}

// The number of rows of data
- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
    return self.selectedTracks.count;
}

// The data to return for the row and component (column) that's being passed in
- (NSString*)pickerView:(UIPickerView *)pickerView titleForRow:(NSInteger)row forComponent:(NSInteger)component {
    return ((Track *)self.selectedTracks[row]).title;
}

// Catpure the picker view selection
- (void)pickerView:(UIPickerView *)pickerView didSelectRow:(NSInteger)row inComponent:(NSInteger)component {
    [self selectTrack:((Track *)self.selectedTracks[row])];
}

/*********************************/
#pragma mark - Actions
/*********************************/
    
- (IBAction)playTouched:(UIButton *)sender {
    if(!self.player.isPlaying) {
        self.playheadTimer = [NSTimer scheduledTimerWithTimeInterval:0.5f target:self selector:@selector(playheadUpdate) userInfo:nil repeats:YES];
        [self.player play];
    }
}
    
- (IBAction)pauseTouched:(UIButton *)sender {
    if(self.player.isPlaying) {
        [self.playheadTimer invalidate];
        self.playheadTimer = nil;
        [self.player pause];
    }
}
    
- (IBAction)playheadValueChanged:(UISlider *)sender {
    NSLog(@"playhead value: %f", sender.value);
    self.player.currentTime = self.player.duration * sender.value;
}
   
- (void)playheadUpdate {
    self.playheadSlider.value = self.player.currentTime / self.player.duration;
}
@end

File: VRSample/VRSample/ViewController.swift
	//
//  ViewController.swift
//  BasicSample-Swift
//
//  Created by Eliza Sapir on 12/04/2017.
//  Copyright © 2017 Kaltura. All rights reserved.
//

import UIKit
import PlayKit
import PlayKitProviders

/*
 This sample will show you how to create a player with basic functionality.
 The steps required:
 1. Load player with plugin config (only if has plugins).
 2. Set delegate.
 3. Prepare Player.
 4. impliment delegate method: shouldAddPlayerViewController.
 5. Get PKVRController.
 6. Use PKVRController API.
 */

struct MediaData {
    let serverURL: String
    let partnerId: Int
    let entryId: String
}
    
class ViewController: UIViewController {
    
    var mediaData: MediaData?
    
    var player: Player?
    var playheadTimer: Timer?
    @IBOutlet weak var playerContainer: PlayerView!
    @IBOutlet weak var playheadSlider: UISlider!
    
    deinit {
        print("deinit")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.playheadSlider.isContinuous = false;
        
        // 1. Load the player
        self.player = PlayKitManager.shared.loadPlayer(pluginConfig: nil)

        // 3. Prepare the player (can be called at a later stage, preparing starts buffering the video)
        self.preparePlayer()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        player?.addObserver(self, event: PlayerEvent.error, block: { (event) in
            print("error: " + (event.error?.localizedDescription ?? ""))
        })
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        self.player?.stop()
        player = nil
        playheadTimer = nil

        player?.removeObserver(self, event: PlayerEvent.error)
        super.viewWillDisappear(animated)
    }
    
/************************/
// MARK: - Player Setup
/***********************/
    func preparePlayer() {
        // setup the player's view
        self.player?.view = self.playerContainer
        
        guard let serverURL = mediaData?.serverURL,
            let partnerId = mediaData?.partnerId,
            let entryID = mediaData?.entryId
            else { return }
        
        let sessionProvider = SimpleSessionProvider(serverURL: serverURL, partnerId: Int64(partnerId), ks: nil)
        let mediaProvider: OVPMediaProvider = OVPMediaProvider(sessionProvider)
        
        mediaProvider.entryId = entryID
        
        mediaProvider.loadMedia { [weak self] (pkMediaEntry, error) in
            guard let mediaEntry = pkMediaEntry else { return }
            
            // create media config
            let mediaConfig = MediaConfig(mediaEntry: mediaEntry)
            self?.player?.prepare(mediaConfig)
        }
    }
    
/************************/
// MARK: - Actions
/***********************/
    
    @IBAction func playTouched(_ sender: Any) {
        guard let player = self.player else {
            print("player is not set")
            return
        }
        
        if !(player.isPlaying) {
            self.playheadTimer = Timer.scheduledTimer(withTimeInterval: 0.5, repeats: true, block: { [weak self] (timer) in
                self?.playheadSlider.value = Float(player.currentTime / player.duration)
            })
            
            player.play()
        }
    }
    
    @IBAction func pauseTouched(_ sender: Any) {
        guard let player = self.player else {
            print("player is not set")
            return
        }
        
        self.playheadTimer?.invalidate()
        self.playheadTimer = nil
        player.pause()
    }
    
    @IBAction func playheadValueChanged(_ sender: Any) {
        guard let player = self.player else {
            print("player is not set")
            return
        }
        
        let slider = sender as! UISlider
        
        print("playhead value:", slider.value)
        player.currentTime = player.duration * Double(slider.value)
    }
    
    @IBOutlet weak var vrBtn: UIButton!

/************************/
// MARK: - VR
/***********************/

    @IBAction func setVRMode(_ sender: Any) {
        // 5. Get PKVRController
        let vrController = self.player?.getController(ofType: PKVRController.self)
        // 6. Use PKVRController API
        vrController?.setVRModeEnabled(true)
    }
}
